### PR TITLE
Updated the module to version 3.1.10

### PR DESCRIPTION
**Pull request by emcAlexandreD**
Sorry for all the files with the same content (-all +all), it seems that some of them were using the wrong encoding.

### CODE REVIEW

File: envoimoinscher.php [hunk 1]
```diff
@@ -995,69 +1017,129 @@ public function ordersTable()
...
-			'perPage' => $per_page
-		));
-		$start = ($page_error - 1) * $per_page;
-		$limits['errors'] = 'LIMIT '.$start.','.$per_page;
-		$smarty->assign('pager_error', $pager_error->setPages());
+		$orders = $this->model->getEligibleOrders($params, $limits);
 
-		$orders = $this->model->getEligibleOrders(array('lang' => $cookie->id_lang), $limits);
+		//get enabled carriers
+		$sql = 'SELECT id_carrier, name FROM '._DB_PREFIX_.'carrier WHERE deleted=0';
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: Maybe you should add a condition to get only your carriers

File: envoimoinscher.php [hunk 2]
```diff
@@ -995,69 +1017,129 @@ public function ordersTable()
...
+				case 'week':
+					$params['filterBy']['start_order_date'] = date('Y-m-d', strtotime('-1 week'));
+					break;
+
+				case 'day':
+					$params['filterBy']['start_order_date'] = date('Y-m-d', strtotime('-1 day'));
+					break;
+			}
+			$smarty->assign('filters', $params['filterBy']);
+		}
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: You could create here the variables ```$sql``` and ```$filter_url```, at the same time. It will avoid the 3 blocks of if/elseif/switch etc

File: envoimoinscher.php [hunk 3]
```diff
@@ -3423,55 +3703,29 @@ public function lookForCarrierUpdates()
...
-			}
-			Db::getInstance()->autoExecute(
-				_DB_PREFIX_.'emc_orders',
-				array('ref_ope_eor' => $ors_class->order_info['opeRef']),
-				'UPDATE',
-				''._DB_PREFIX_.'orders_id_order = '.(int)Tools::getValue('order')
-			);
-			if (EnvoimoinscherModel::TRACK_OPE_TYPE == $config['EMC_TRACK_MODE'])
+			$documents = Db::getInstance()->ExecuteS('SELECT * FROM '._DB_PREFIX_.'emc_documents
+				WHERE generated_ed = 1 AND '._DB_PREFIX_.'orders_id_order in ('.implode(',', $orders_id).')');
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: You could protect your data with ```array_map('intval', $orders_id)```

File: envoimoinscher.php [hunk 4]
```diff
@@ -3423,55 +3703,29 @@ public function lookForCarrierUpdates()
...
-			'user' => $config['EMC_LOGIN'],
-			'pass' =>	$config['EMC_PASS'],
-			'key' => $config['EMC_KEY']));
-		$ors_class->setPlatformParams($this->ws_name, _PS_VERSION_, $this->version);
-		$ors_class->setEnv(Tools::strtolower($config['EMC_ENV']));
-		$ors_class->setGetParams(array('module' => $this->ws_name, 'version' => $this->local_version));
-		$ors_class->getOrderInformations(Tools::getValue('ref'));
-		ob_end_clean();
-		if (!$ors_class->curl_error)
+		$emc = new Envoimoinscher();
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: In this file, you can replace ```$emc``` with ```$this``` and remove this line.

File: envoimoinscher.php [hunk 5]
```diff
@@ -3256,29 +3512,49 @@ public function downloadLabels()
...
 			foreach ($references as $reference)
 				$refs[] = $reference['ref_emc_eor'];
 		}
 
+		// We now query the document
 		if (count($refs) > 0)
 		{
-			$slip_url = stristr($references[0]['link_ed'], $references[0]['ref_emc_eor'], true);
+			// Get the document's base url
+			$base_url = explode('?', $references[0]['link_ed'])[0];
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: This statement is incompatible with PHP 5.1/5.2. You should write these lines:
```php
$base_url = explode('?', $references[0]['link_ed']);
$base_url = $base_url[0];
```

File: includes/EnvoimoinscherModel.php [hunk 6]
```diff
@@ -278,26 +278,135 @@ public function getEligibleOrdersCount()
...
+			if (isset($params['filterBy']['filter_id_order']))
+				$sql .= ' AND o.id_order = '.(int)$params['filterBy']['filter_id_order'];
+
+			//by order status
+			if (isset($params['filterBy']['status']) && is_array($params['filterBy']['status']))
+			{
+				$filter = ' AND o.current_state IN (';
+				$statuses = array();
+				foreach ($params['filterBy']['status'] as $value)
+					array_push($statuses, $value);
```
Comment (Quetzacoalt91) [inline includes/EnvoimoinscherModel.php]:
```php
$statuses[] = (int)$value;
```
Safer with the int cast

File: includes/EnvoimoinscherModel.php [hunk 7]
```diff
@@ -278,26 +278,135 @@ public function getEligibleOrdersCount()
...
+				$filter .= implode(',', $statuses).')';
+				if (count($statuses) > 0)
+					$sql .= $filter;
```
Comment (Quetzacoalt91) [inline includes/EnvoimoinscherModel.php]:
```php
	if (count($statuses) > 0)
	{
		$filter .= implode(',', $statuses).')';
		$sql .= $filter;
	}
```

I add this line in the test, because its execution is useless if ```$statuses``` is empty.

File: includes/EnvoimoinscherModel.php [hunk 8]
```diff
@@ -278,26 +278,135 @@ public function getEligibleOrdersCount()
...
+				case '2':
+					$sql .= ' AND er.errors_eoe != ""';
+					break;
+				default:
+					break;
+			}
+
+			//add default order status filter chosen in module config pages
+			$filter = ' AND o.current_state IN (';
+			$filter .= implode(',', explode(';', Configuration::get('EMC_FILTER_STATUS'))).')';
```
Comment (Quetzacoalt91) [inline includes/EnvoimoinscherModel.php]: Each data it this array should be escaped, with array_map(), pSQL or intval.

File: includes/EnvoimoinscherModel.php [hunk 9]
```diff
@@ -264,13 +264,13 @@ public function getDimensions()
 	}
 
 	/**
-	* Gets the number of EnvoiMoinsCher orders and other orders which haven't been send yet.
+	* Gets the number of EnvoiMoinsCher orders and other orders which haven't been sent yet.
 	* @access public
 	* @return int Orders count.
 	*/
-	public function getEligibleOrdersCount()
+	public function getEligibleOrdersCount($params)
```
Comment (Quetzacoalt91) [inline includes/EnvoimoinscherModel.php]: Actually, This function could be really more simple !

```php
public function getEligibleOrdersCount($params)
{
	return $this->getEligibleOrders($params);
}
```

File: includes/EnvoimoinscherModel.php [hunk 10]
```diff
@@ -1632,4 +1856,193 @@ public static function getNameCategory($id_eca)
...
+		 WHERE eo.'._DB_PREFIX_.'orders_id_order = '.$order.' AND eo.tracking_eor = "'.$key.'" ')) > 0;
+	}
+
+	public function handlePush()
+	{
+		$type 		= Tools::getValue('type');
+		$key			= Tools::getValue('key');
+		$order 		= (int)Tools::getValue('order');
+		$return 	= false;
+		$emc 			= new Envoimoinscher();
```
Comment (Quetzacoalt91) [inline includes/EnvoimoinscherModel.php]:
```php
$emc = Module::getInstanceByName('envoimoinscher');
```

would be better, because it uses a cache system.

File: includes/EnvoimoinscherModel.php [hunk 11]
```diff
@@ -1632,4 +1856,193 @@ public static function getNameCategory($id_eca)
...
+		}
+
+		// Generate the default date if no one was given
+		if ($date == false)
+			$date = time();
+
+		// Add tracking info
+		Db::getInstance()->Execute('INSERT INTO '._DB_PREFIX_.'emc_tracking
+			 ('._DB_PREFIX_.'orders_id_order, state_et, date_et, text_et, localisation_et)
+			 VALUES ('.$order.', "'.$state.'", "'.date('Y-m-d H:i:s', $date).'", "'.pSQL($text).'", "'.pSQL($localisation).'")');
```
Comment (Quetzacoalt91) [inline includes/EnvoimoinscherModel.php]: It would be good to escape the data inserted into the database with pSQL or an int cast.

File: sql/install.sql [hunk 12]
```diff
@@ -1,367 +1,367 @@
...
+-- REQUEST --
+DROP TABLE IF EXISTS `{PREFIXE}emc_documents`;
+-- REQUEST --
+CREATE TABLE IF NOT EXISTS `{PREFIXE}emc_documents` (
+	`id_ed` int(11) NOT NULL AUTO_INCREMENT,
+	`{PREFIXE}orders_id_order` int(10) unsigned NOT NULL,
+	`{PREFIXE}cart_id_cart` int(10) unsigned NOT NULL,
+	`link_ed` varchar(255) NOT NULL,
+	`generated_ed` int(1) NOT NULL DEFAULT 0,
+	`type_ed` varchar(100) NOT NULL,
```
Comment (Quetzacoalt91) [inline sql/install.sql]: Because the database structure has been modified, you have to create an upgrade file which will alter your table structure during the upgrade.

**Comment by Quetzacoalt91** [comment]
During the step number 3 of the first setup, I get the following notices.

capture du 2015-01-09 15 02 37

(It seems that my credentials are invalid at that time).

**Comment by Quetzacoalt91** [comment]
I get the following error when I want to get the previous orders:
capture du 2015-01-09 15 14 12

**Comment by Quetzacoalt91** [comment]
When I click on the button "All orders" on the "Orders pending shipment" page, I get the following result:
capture du 2015-01-09 15 19 10

**Comment by Quetzacoalt91** [comment]
Making an quotation does not recover properly from unexpected behavior:
capture du 2015-01-09 15 27 44

File: Env/CarriersList.php [hunk 13]
```diff
@@ -57,8 +57,8 @@ class Env_CarriersList extends Env_WebService
 	 */
 	public function getCarriersList($channel, $version)
 	{
-		$this->param['channel'] = Tools::strtolower($channel);
-		$this->param['version'] = Tools::strtolower($version);
+		$this->param['channel'] = strtolower($channel);
+		$this->param['version'] = strtolower($version);
```
Comment (emcAlexandreD) [inline Env/CarriersList.php]: We want to avoid any module's code in our library (all the files in the Env folder) but we can still add it back if this is a problem for you

File: envoimoinscher.php [hunk 14]
```diff
@@ -161,15 +161,17 @@ public function install()
...
 			Logger::addLog($error);
 			return false;
 		}
 
 		// Get default configuration state
 		$states = OrderState::getOrderStates((int)$this->getContext()->language->id);
+		$states_array = array();
 		foreach ($states as $state)
 		{
+					array_push($states_array, $state['id_order_state']);
```
Comment (emcAlexandreD) [inline envoimoinscher.php]: Done with the incoming revision

File: envoimoinscher.php (hunk 3 above)
Comment (emcAlexandreD) [inline envoimoinscher.php]: Done with the incoming revision

File: envoimoinscher.php (hunk 4 above)
Comment (emcAlexandreD) [inline envoimoinscher.php]: Done with the incoming revision

File: envoimoinscher.php (hunk 5 above)
Comment (emcAlexandreD) [inline envoimoinscher.php]: Done with the incoming revision

File: includes/EnvoimoinscherModel.php (hunk 6 above)
Comment (emcAlexandreD) [inline includes/EnvoimoinscherModel.php]: Done with the incoming revision

File: includes/EnvoimoinscherModel.php (hunk 7 above)
Comment (emcAlexandreD) [inline includes/EnvoimoinscherModel.php]: Done with the incoming revision

File: includes/EnvoimoinscherModel.php (hunk 8 above)
Comment (emcAlexandreD) [inline includes/EnvoimoinscherModel.php]: Done with the incoming revision

File: includes/EnvoimoinscherModel.php (hunk 9 above)
Comment (emcAlexandreD) [inline includes/EnvoimoinscherModel.php]: Done with the incoming revision

File: includes/EnvoimoinscherModel.php (hunk 10 above)
Comment (emcAlexandreD) [inline includes/EnvoimoinscherModel.php]: Done in the next revision here and on the other instances in EnvoimoinscherModel and EnvoimoinscherOrder.

File: includes/EnvoimoinscherModel.php (hunk 11 above)
Comment (emcAlexandreD) [inline includes/EnvoimoinscherModel.php]: Done with the incoming revision

File: envoimoinscher.php (hunk 2 above)
Comment (emcAlexandreD) [inline envoimoinscher.php]: Done with the incoming revision for $filter_url. For the $sql part, even if your solution is faster, i think it is easier to understand the code if the $sql "filter part" is built in getEligibleOrders function

File: sql/install.sql (hunk 12 above)
Comment (emcAlexandreD) [inline sql/install.sql]: There are some files missing in the last commit, including the Upgrade file, my apologies, it will be added with the next revision

**Comment by emcAlexandreD** [comment]
These warnings come from our library, i will just hide the warnings for now and remove them when it will be fixed in the library.

**Comment by emcAlexandreD** [comment]
I just pushed a correction for these feedback, they should be now fixed

File: Env/CarriersList.php (hunk 13 above)
Comment (Quetzacoalt91) [inline Env/CarriersList.php]: So the folder *Env* is a library ... Okay, leave it untouched then.

**Comment by Quetzacoalt91** [comment]
Hi,

I am sorry but I see only new files in your last commit and none of my comments became "outdated". Is everything commited ?

**Comment by emcAlexandreD** [comment]
You're right, 3 files were missing, i just pushed them. (and i thought i was now familiar with git).

File: envoimoinscher.php [hunk 15]
```diff
@@ -2811,7 +3041,7 @@ public function makeApiCall($params)
 		$offers_quotation = array();
 		if (count($offers_db) == 1) // correction needed for quotation
 		{
-			$cot_cl->getQuotation($quot_info);
+			@$cot_cl->getQuotation($quot_info);
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: Surrounding with a try/catch should be less dirty than a ```@``` and will allow you to manage an error properly.

File: envoimoinscher.php [hunk 16]
```diff
@@ -2824,7 +3054,7 @@ public function makeApiCall($params)
 				$cot_cl->setParamMulti($quot_info);
 			}
 
-			$cot_cl->getQuotationMulti();
+			@$cot_cl->getQuotationMulti();
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: Same comment with the try/catch

File: upgrade/Upgrade-3.1.10.php [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+ */
+
+function upgrade_module_3_1_10($module)
+{
+	//Set default configuration
+	Configuration::updateValue('EMC_ENABLED_LOGS', 0);
+	$states = OrderState::getOrderStates((int)$module->getContext()->language->id);
+	$states_array = array();
+	foreach ($states as $state)
+		array_push($states_array, $state['id_order_state']);
```
Comment (Quetzacoalt91) [inline upgrade/Upgrade-3.1.10.php]: As I said on the first validation of this PR, using ```$states_array[] = $state['id_order_state'];``` is better if you need to add only one value.

**Comment by Quetzacoalt91** [comment]
I have a small request about your Google Maps API implementation. I use this PR to avoid a new one later.
A bug has been reporting when a merchant uses different carrier modules on the same shop. We detected a conflict between them when they both use Gmaps API. Can you please include ```//maps.google.com/maps/api/js?sensor=false``` with the addJs function ? Here is an example: https://github.com/PrestaShop/tntcarrier/blob/eca379f0d6618a99e18239af9631cf9e96af8af9/tntcarrier.php#L981

Thanks

File: envoimoinscher.php (hunk 15 above)
Comment (emcAlexandreD) [inline envoimoinscher.php]: The problem is that we cannot catch warnings with a try catch, the only way i found is to use set_error_handler or to set warnings as errors, but this is not something we are supposed to change. Maybe there is something i missed?
The warnings you saw was due to a problem with the library when it receive a malformed response (like a 404 or 500 error) so this is something we will fix in the library and then add to the module. After this we will remove the @.

Comment (Quetzacoalt91) [inline envoimoinscher.php]: Okay, I made a mistake between cURL and SOAP calls.

Anyway, you should add to have a better management of errors:
```php
	$result = curl_exec($req);
	// Check if something went wrong
	if ($result === false)
		[...]
```

File: envoimoinscher.php [hunk 18]
```diff
@@ -2473,6 +2708,7 @@ public function hookupdateCarrier($params)
 	 */
 	public function hookHeader()
 	{
+		$this->getContext()->controller->addJs('https://maps.google.com/maps/api/js?sensor=false');
```
Comment (Quetzacoalt91) [inline envoimoinscher.php]: Is it possible to have it in the hook hookDisplayCarrierList ? It will be better to load it only when needed.

Comment (emcAlexandreD) [inline envoimoinscher.php]: My bad, i thought we had to add the js/css in the header hook only, i'll change it now

**Comment by Quetzacoalt91** [comment]
Hi @emcAlexandreD,

Thanks for your PR, it has just been merged. For the next time, it would be good to remove the ```@``` added before ```getQuotation()```. As I said, you should check i.e that ```curl_exec``` does not return false.

Regards

**Comment by emcAlexandreD** [comment]
thanks for your time, we will update our library to handle this, once it will be done, the module will receive this new version.
